packages/protobuf/src/equals.ts

```diff
@@ -135,3 +193,91 @@ function fieldEquals(
   }
   return true;
 }
+
+function anyUnpackedEquals(a: Any, b: Any, opts: EqualsOptions): boolean {
+  if (a.typeUrl !== b.typeUrl) {
+    return false;
+  }
+  const unpackedA = anyUnpack(a, opts.registry);
+  const unpackedB = anyUnpack(b, opts.registry);
+  if (unpackedA && unpackedB) {
+    const schema = opts.registry.getMessage(unpackedA.$typeName);
+    if (schema) {
+      return equals(schema, unpackedA, unpackedB, opts);
+    }
+  }
+  return scalarEquals(ScalarType.BYTES, a.value, b.value);
+}
+
+function unknownEquals(
+  a: ReflectMessage,
+  b: ReflectMessage,
+  registry: Registry | undefined,
+) {
+  function getTrulyUnknown(
+    msg: ReflectMessage,
+    registry: Registry | undefined,
+  ) {
+    const u = msg.getUnknown() ?? [];
+    return registry
+      ? u.filter((uf) => !registry.getExtensionFor(msg.desc, uf.no))
+      : u;
+  }
+  const unknownA = getTrulyUnknown(a, registry);
+  const unknownB = getTrulyUnknown(b, registry);
+  if (unknownA.length != unknownB.length) {
+    return false;
+  }
+  for (let i = 0; i < unknownA.length; i++) {
+    const a = unknownA[i],
+      b = unknownB[i];
+    if (a.no != b.no) {
+      return false;
+    }
+    if (a.wireType != b.wireType) {
+      return false;
+    }
+    if (!scalarEquals(ScalarType.BYTES, a.data, b.data)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function extensionsEquals(
+  a: ReflectMessage,
+  b: ReflectMessage,
+  opts: EqualsOptions,
+) {
+  function getSetExtensions(
+    msg: ReflectMessage,
+    registry: Registry,
+  ): DescExtension[] {
+    return (msg.getUnknown() ?? [])
+      .map((uf) => registry.getExtensionFor(msg.desc, uf.no))
+      .filter((e) => !!e)
+      .filter((e, index, arr) => arr.indexOf(e) === index);
```

Comment:
This expression seems like a tautology. What is this trying to filter out? Is this just removing duplicates (where `arr.indexOf(e)` might return an earlier index)?

Reply:
Yes, it's removing duplicates. This expression is widely used, for better or worse.